### PR TITLE
Text generation improvement (UI client, data parallel support)

### PR DESCRIPTION
**Pull request by yidong72**
# What does this PR do ?

1. Added Text generation web server for GPT model and universal prompt encoder.
2. Added Retrieval text generation services and web server.
3. Fixed various bugs for text generation
4. Allow data parallel for text generation.

### CODE REVIEW

File: nemo/collections/nlp/modules/common/megatron/retrieval_service.py [hunk 1]
```diff
@@ -13,64 +13,174 @@
...
+        """
+        raise NotImplementedError()
 
-class FaissRetrievalService(RetrievalService):
+
+class ChunkStore:
+    def __init__(self, chunk_size, pad_id):
+        self.store = {}
+        self._count = 0
+        self.no_retrieval = np.ones(2 * chunk_size, dtype=np.int64) * pad_id
```
Comment (soumye) [inline nemo/collections/nlp/modules/common/megatron/retrieval_service.py:71]: What is the use case of having this `self.no_retrieval`? I can see that it ensures it always has atleast one `dummy` element

File: nemo/collections/nlp/modules/common/megatron/retrieval_service.py [hunk 2]
```diff
@@ -13,64 +13,174 @@
...
+        self.tokenizer = tokenizer
+        self.pool = pool
+        self.sentence_bert_batch = sentence_bert_batch
+        self.embedding_dim = self.bert_model.get_sentence_embedding_dimension()
+
+    def put(self):
+        # logging.info("request IP: " + str(request.remote_addr))
+        # logging.info(json.dumps(request.get_json()))
+        data = request.get_json()
+        if isinstance(data, dict):
```
Comment (soumye) [inline nemo/collections/nlp/modules/common/megatron/retrieval_service.py:102]: Why do you have this check here? The same doesn't exist in `FaissRetrievalResource`

File: nemo/collections/nlp/modules/common/megatron/retrieval_service.py [hunk 3]
```diff
@@ -97,3 +207,272 @@ def get_knn(self, query: Union[List[str], str, torch.Tensor]):
...
+        api = Api(self.app)
+        api.add_resource(
+            FaissRetrievalResource, '/knn', resource_class_args=[self.index, self.tokenizer, self.ds,],
+        )
+
+    def run(self, url, port=PORT_NUM):
+        threading.Thread(target=lambda: self.app.run(host=url, threaded=True, port=port)).start()
+
+
+class DynamicRetrievalResource(FaissRetrievalResource):
```
Comment (soumye) [inline nemo/collections/nlp/modules/common/megatron/retrieval_service.py:247]: What's the use case for this over plain `RetrievalResource`? A few comments would be helpful here imo.

File: nemo/collections/nlp/modules/common/megatron/retrieval_service.py [hunk 4]
```diff
@@ -97,3 +207,272 @@ def get_knn(self, query: Union[List[str], str, torch.Tensor]):
...
+                token_ids.append(self.tokenizer.eos_id)
+            np_array = np.array(token_ids, dtype=np.int32)
+            padded_size = self.chunk_size - (len(np_array) % self.chunk_size)
+            # for retrieval database, added one more chunk in the end as padding
+            padded_size += self.chunk_size
+            np_array = np.pad(np_array, (0, padded_size), 'constant', constant_values=self.pad_id)
+            chunk_texts = []
+            for i in range(0, len(np_array), self.stride):
+                if i + 2 * self.chunk_size <= len(np_array):
+                    chunk = np_array[i : i + 2 * self.chunk_size]
```
Comment (soumye) [inline nemo/collections/nlp/modules/common/megatron/retrieval_service.py:303]: why is a chunk of size `2 * self.chunk_size`? A chunk is `64` tokens only right?

File: nemo/collections/nlp/modules/common/megatron/retrieval_service.py [hunk 5]
```diff
@@ -97,3 +207,272 @@ def get_knn(self, query: Union[List[str], str, torch.Tensor]):
...
+            sentence_list = []
+            for q in query:
+                text = self.tokenizer.ids_to_text(q)
+                sentence_list.append(text)
+            query = sentence_list
+        data = {'sentences': query, 'add_eos': add_eos}
+        return request_data(data, PORT_NUM_DYN)
+
+
+class ComboRetrievalService(RetrievalService):
```
Comment (soumye) [inline nemo/collections/nlp/modules/common/megatron/retrieval_service.py:434]: Another comment here would be helpful. In general the differences of use cases.

File: nemo/collections/nlp/modules/common/text_generation_server.py [hunk 6]
```diff
@@ -142,11 +144,36 @@ def put(self):
             if min_tokens_to_generate < 0:
                 return "min_tokens_to_generate must be an integer no less than 0"
 
+        neighbors = None
+        if "neighbors" in request.get_json():
+            neighbors = request.get_json()["neighbors"]
+            if not isinstance(neighbors, int):
+                return "num of neighbors must be an integer no less than 0"
+            if neighbors < 0:
```
Comment (soumye) [inline nemo/collections/nlp/modules/common/text_generation_server.py:152]: is `neighbors=0` an acceptable number too? Would you not just ignore then?

File: nemo/collections/nlp/modules/common/text_generation_strategy.py [hunk 7]
```diff
@@ -280,18 +285,118 @@ def __init__(self, model, **args):
...
+        convert the sentences into lists of tokens, pad them to the same length, add bos tokens if it is needed
+        Args:
+            sentences (List[str]): list of input sentences in str format.
+            max_len (int): max number of tokens to generate.
+            add_BOS (bool): whether to add the BOS token at the beginning
+        Returns:
+            Tuple[torch.Tensor], the tokenized and padded torch tensor and the token context length tensor.
+        """
+        tokenizer = self.model.tokenizer
+        if add_BOS:
```
Comment (soumye) [inline nemo/collections/nlp/modules/common/text_generation_strategy.py:330]: For `add_BOS` you are adding `tokenizer.eos_id` here. Is this corrrect?

File: nemo/collections/nlp/modules/common/text_generation_strategy.py [hunk 8]
```diff
@@ -347,7 +458,11 @@ def prepare_batch_at_step(
             [set_inference_key_value_memory] * micro_batch_size, device=torch.cuda.current_device()
         )
         len_array = torch.tensor([maxlen] * micro_batch_size, device=torch.cuda.current_device())
-        neighbors_array = torch.tensor([self.neighbors] * micro_batch_size, device=torch.cuda.current_device())
+        if self.neighbors == 0:
+            # no retrieval, use 1 padding
```
Comment (soumye) [inline nemo/collections/nlp/modules/common/text_generation_strategy.py:462]: Why do we need to do this?

**Review by soumye — DISMISSED** [review]
Great and helpful PR! I have left some comments for some clarification. I am not well-versed with web-server or communication related parts but apart from that LGTM.

File: nemo/collections/nlp/modules/common/megatron/mup/layer.py [hunk 9]
```diff
@@ -71,13 +71,16 @@ def __init__(self, mpu_vocab_size, parallel_output):
...
+        self.warn_once = False
 
     def forward(self, hidden_states, word_embeddings_weight):
         if hasattr(word_embeddings_weight, 'infshape'):
             width_mult = word_embeddings_weight.infshape.width_mult()
         else:
             width_mult = 1.0
-            logging.warning("need to set_shape before use mu-Transfer readout layer")
+            if not self.warn_once:
+                logging.warning("need to set_shape before use mu-Transfer readout layer")
```
[the remark this answers is not in the document]
Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/mup/layer.py:82]: This is related to muTransfer. If the mu transfer shape file is not provided, it only warns once. So this layer can work both in muTransfer on and off cases.

File: nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py [hunk 10]
```diff
@@ -340,6 +340,8 @@ def validation_step(self, batch, batch_idx):
         return reduced_loss
 
     def validation_epoch_end(self, outputs):
+        if len(outputs) == 0:
```
[the remark this answers is not in the document]
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py:343]: Yes. sometimes the outputs is empty. This is a work-around to prevent it crash.

File: scripts/nlp_language_modeling/build_retrieval_index.py [hunk 11]
```diff
@@ -307,8 +307,18 @@ def get_emb():
...
+        if args.devices is None or not torch.cuda.is_available():
+            device_list = None
+        else:
+            device_list = ['cuda:' + str(device) for device in args.devices.split(',')]
+        index = faiss.read_index(args.output_file)
+        co = faiss.GpuMultipleClonerOptions()
+        co.useFloat16 = True
+        co.usePrecomputed = False
+        co.shard = True
+        index = faiss.index_cpu_to_all_gpus(index, co, ngpu=len(device_list))
```
[the remark this answers is not in the document]
Comment (yidong72) [reply scripts/nlp_language_modeling/build_retrieval_index.py:319]: When computing the distributed Faiss index, it creates an extra companion large merge file in the disk to work with the index,. By loading it to GPU as distributed Faiss index and then convert it CPU and dump to disk. The merge file will be included in the Faiss index, which is easy to maintain.

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 12]
```diff
@@ -2259,7 +2259,7 @@ def get_num_layers(self, num_layers):
                     num_layers = num_layers // num_ranks_in_encoder
                 else:
                     num_layers = num_layers // num_ranks_in_decoder
-            else:
+            elif self.model_type == ModelType.encoder_or_decoder:
```
[the remark this answers is not in the document]
Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:2262]: yes. it can be set to None.

File: examples/nlp/language_modeling/megatron_retro_eval.py [hunk 13]
```diff
@@ -86,26 +97,55 @@ def main(cfg) -> None:
...
-    if not cfg.use_predict_method:
-        # First method of running text generation, call model.generate method
-        response = model.generate(
-            inputs=OmegaConf.to_container(cfg.prompts),
-            length_params=length_params,
-            sampling_params=sampling_params,
-            **cfg.retrieval_service,
-        )
+    # check whether the DDP is initialized
+    if parallel_state.is_unitialized():
```
[the remark this answers is not in the document]
Comment (yidong72) [reply examples/nlp/language_modeling/megatron_retro_eval.py:101]: Megatron models support model parallelism. To run evaluation, we need to spawn all the necessary processes to do the inference. Here is to check whether the parallel_state is not initialized, then we can manually spawn all the workers.

File: nemo/collections/nlp/modules/common/megatron/retrieval_service.py (hunk 4 above)
Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/retrieval_service.py:303]: Query chunk has 64 tokens. this is KNN context chunk, which has 128 tokens.

File: nemo/collections/nlp/modules/common/megatron/retrieval_service.py (hunk 5 above)
Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/retrieval_service.py:434]: added comments.

File: nemo/collections/nlp/modules/common/text_generation_server.py (hunk 6 above)
Comment (yidong72) [reply nemo/collections/nlp/modules/common/text_generation_server.py:152]: yes. neighbor = 0 is acceptable. In the case of K=0, we will use padding as context token, and the attention mask will mask them out.

File: nemo/collections/nlp/modules/common/text_generation_strategy.py (hunk 7 above)
Comment (yidong72) [reply nemo/collections/nlp/modules/common/text_generation_strategy.py:330]: yeah. the GPT tokenizer uses "eos" as "bos". tokenizer doesn't distinguish them. But you are right, bos_id is better here in case other tokenizer uses a different id. fixed

File: scripts/nlp_language_modeling/build_retrieval_index.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (soumye) [reply scripts/nlp_language_modeling/build_retrieval_index.py:319]: I see.

File: nemo/collections/nlp/modules/common/megatron/retrieval_service.py (hunk 4 above)
Comment (soumye) [reply nemo/collections/nlp/modules/common/megatron/retrieval_service.py:336]: Ah. I see.

**Review by soumye — APPROVED** [review]
LGTM. Thanks for the clarifications.
